src/framework/constraints.ml

```diff
@@ -549,37 +587,36 @@ struct
 
   let startcontext () = Some (S.startcontext ())
   let name () = S.name ()^" with context gas"
-  let startstate v = S.startstate v, get_int "ana.context.gas_value"
-  let exitstate v = S.exitstate v, get_int "ana.context.gas_value"
+  let startstate v = S.startstate v, Gas.startgas ()
+  let exitstate v = S.exitstate v, Gas.startgas ()
   let morphstate v (d,i) = S.morphstate v d, i
 
   let conv (ctx:(D.t,G.t,C.t,V.t) ctx): (S.D.t,G.t,S.C.t,V.t)ctx =
-    if (cg_val ctx <= 0)
-    then {ctx with local = fst ctx.local
-                 ; split = (fun d es -> ctx.split (d, cg_val ctx) es)
-                 ; context = (fun () -> ctx_failwith "no context (contextGas = 0)")}
-    else {ctx with local = fst ctx.local
-                 ; split = (fun d es -> ctx.split (d, cg_val ctx) es)
-                 ; context = (fun () -> Option.get (ctx.context ()))}
+    {ctx with local = fst ctx.local
+            ; split = (fun d es -> ctx.split (d, cg_val ctx) es)
+            ; context = (fun () -> match ctx.context () with Some c -> c | None -> ctx_failwith "no context (contextGas = 0)")}
 
   let context ctx fd (d,i) =
     (* only keep context if the context gas is greater zero *)
-    if i <= 0 then None else Some (S.context (conv ctx) fd d)
+    if Gas.is_exhausted fd i then
+      None
+    else
+      Some (S.context (conv ctx) fd d)
 
   let enter ctx r f args =
-    (* callee gas = caller gas - 1 *)
-    let liftmap_tup = List.map (fun (x,y) -> (x, cg_val ctx), (y, max 0 (cg_val ctx - 1))) in
+    let liftmap_tup = List.map (fun (x,y) -> (x, cg_val ctx), (y, Gas.callee_gas f (cg_val ctx))) in
     liftmap_tup (S.enter (conv ctx) r f args)
 
   let threadenter ctx ~multiple lval f args =
-    let liftmap f = List.map (fun (x) -> (x, max 0 (cg_val ctx - 1))) f in
+    let liftmap d = List.map (fun (x) -> (x, Gas.thread_gas f (cg_val ctx))) d in
     liftmap (S.threadenter (conv ctx) ~multiple lval f args)
 
   let query ctx (type a) (q: a Queries.t):a Queries.result =
     match q with
     | Queries.GasExhausted ->
+      (* The query is only used in a way where overapproximating gas exhaustion is not harmful *)
       let (d,i) = ctx.local in
-      (i <= 0)
+      Gas.is_any_exhausted i
```

Comment:
The need for this query is explained here: https://github.com/goblint/analyzer/pull/1508 and https://github.com/goblint/analyzer/pull/1475.

The idea is to avoid calls to `sync` for start nodes of functions when the `threadflag` is context-sensitive, i.e., the abstract value at a start point is never given as the join of things in ST and MT mode (which requires handling similar to entering multi-threaded mode to be sound for some privatizations). Normally, this can be decided just by whether `threadflag` is not configured to be context-insensitive, but the gas may make this analysis context-insensitive even if it is not in general, which is why this query is needed.

The place where this query is asked is here:

https://github.com/goblint/analyzer/blob/3c0ff8c927558a574c7454c6d08d9e8b445d1ebe/src/analyses/commonPriv.ml#L63-L80

So it is asked when the threadflag is context sensitive. The `Gas.is_any_exhausted` is only called when the lifter is actually active.

https://github.com/goblint/analyzer/blob/3c0ff8c927558a574c7454c6d08d9e8b445d1ebe/src/analyses/mCP.ml#L321-L327


This function is called by 3 relational and 3 non-relational privatizations when a sync edge with reason `JoinCall` is encountered to determine whether its handling of `branched_sync` needs to be called.

https://github.com/goblint/analyzer/blob/3c0ff8c927558a574c7454c6d08d9e8b445d1ebe/src/analyses/basePriv.ml#L433-L437


That call originates here:

https://github.com/goblint/analyzer/blob/3c0ff8c927558a574c7454c6d08d9e8b445d1ebe/src/framework/constraints.ml#L668-L674

It is thus called once per function call.

Reply:
The comments are still true, but by giving `JoinCall` the `fundec` as an argument and also giving the query the argument, no over-approximation is needed anymore!